pkg/ebpf/uprobes/attacher.go

```diff
@@ -0,0 +1,721 @@
+	// ErrNoMatchingRule is returned when no rule matches the shared library path.
+	ErrNoMatchingRule = errors.New("no matching rule")
+	// regex that defines internal DataDog processes
+	internalProcessRegex = regexp.MustCompile("datadog-agent/.*/((process|security|trace)-agent|system-probe|agent)")
+)
+
+// AttachTarget defines the target to which we should attach the probes, libraries or executables
+type AttachTarget uint8
+
+const (
+	// AttachToExecutable attaches to the main executable
+	AttachToExecutable AttachTarget = 1 << iota
+	// AttachToSharedLibraries attaches to shared libraries
+	AttachToSharedLibraries
+)
+
+// AttachRule defines how to attach a certain set of probes. Uprobes can be attached
+// to shared libraries or executables, this structure tells the attacher which ones to
+// select and to which targets to do it.
+type AttachRule struct {
+	// LibraryNameRegex defines which libraries should be matched by this rule
+	LibraryNameRegex *regexp.Regexp
+	// Targets defines the targets to which we should attach the probes, shared libraries and/or executables
+	Targets AttachTarget
+	// ProbesSelectors defines which probes should be attached and how should we validate
+	// the attachment (e.g., whether we need all probes active or just one of them, or in a best-effort basis)
+	ProbesSelector []manager.ProbesSelector
+}
+
+func (r *AttachRule) canTarget(target AttachTarget) bool {
+	return r.Targets&target != 0
+}
+
+func (r *AttachRule) matchesLibrary(path string) bool {
+	return r.canTarget(AttachToSharedLibraries) && r.LibraryNameRegex != nil && r.LibraryNameRegex.MatchString(path)
+}
+
+func (r *AttachRule) matchesExecutable(_ string) bool {
+	return r.canTarget(AttachToExecutable)
+}
+
+// AttacherConfig defines the configuration for the attacher
+type AttacherConfig struct {
+	// Rules defines a series of rules that tell the attacher how to attach the probes
+	Rules []*AttachRule
+
+	// ScanTerminatedProcessesInterval defines the interval at which we scan for terminated processes. Set
+	// to zero to disable
+	ScanTerminatedProcessesInterval time.Duration
+
+	// ProcRoot is the root directory of the proc filesystem
+	ProcRoot string
+
+	// ExcludeTargets defines the targets that should be excluded from the attacher
+	ExcludeTargets ExcludeMode
+
+	// EbpfConfig is the configuration for the eBPF program
+	EbpfConfig *ebpf.Config
+
+	// PerformInitialScan defines if the attacher should perform an initial scan of the processes before starting the monitor
+	PerformInitialScan bool
+
+	// ProcessMonitorEventStream defines whether the process monitor is using the event stream
+	ProcessMonitorEventStream bool
+}
+
+// SetDefaults configures the AttacherConfig with default values for those fields for which the compiler
+// defaults are not enough
+func (ac *AttacherConfig) SetDefaults() {
+	if ac.ScanTerminatedProcessesInterval == 0 {
+		ac.ScanTerminatedProcessesInterval = 30 * time.Minute
```

Comment:
30 minutes as default? Isn't that "too slow"?

Reply:
It is, I made a mistake copying the existing default. Changed to 30 second as it's in the USM code.